### PR TITLE
use alignment to distinguish Str from other values (fixes #592)

### LINKED ISSUE #592: admin: improve JSON rendering in admin GUI
Now that we have two ways of formatting JSON result, "readable" and "precise", we should support options to select the output format in the admin UI and in the REPL. At the moment, both are hard-coded to produce the "readable" format, but there is code to support either (just no interface).

In the Admin UI, this would probably take the form of a checkbox that appears with the results table, and immediately re-renders the contents.

In the REPL, it would probably be a command, e.g.`set format = precise`.

### PR DESCRIPTION
**Pull request by mossprescott**
In the admin UI, a custom table renderer now shows Strings as left-aligned,
all other values right-aligned, and in blue.

In the REPL, results are now rendered in an MarkDown-style table using the
same approach to alignment.

Here's a sample:

```
💪 $ select * from zips
Mongo
db.zips.find();


Query time: 0.0s
 city          | loc.0       | loc.1      | pop    | state |
---------------|-------------|------------|--------|-------|
 AGAWAM        |  -72.622739 |  42.070206 |  15338 | MA    |
 CUSHMAN       |   -72.51565 |  42.377017 |  36963 | MA    |
 BARRE         |  -72.108354 |  42.409698 |   4546 | MA    |
 BELCHERTOWN   |  -72.410953 |  42.275103 |  10579 | MA    |
 BLANDFORD     |  -72.936114 |  42.182949 |   1240 | MA    |
 BRIMFIELD     |  -72.188455 |  42.116543 |   3706 | MA    |
 CHESTER       |  -72.988761 |  42.279421 |   1688 | MA    |
 CHESTERFIELD  |  -72.833309 |   42.38167 |    177 | MA    |
 CHICOPEE      |  -72.607962 |  42.162046 |  23396 | MA    |
 CHICOPEE      |  -72.576142 |  42.176443 |  31495 | MA    |
...
```

### CODE REVIEW

File: core/src/main/scala/slamdata/engine/repl/prettify.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import scalaz._
+import Scalaz._
+
+import slamdata.engine._
+import slamdata.engine.fp._
+
+object Prettify {
+  case class Path(segs: List[String]) {
+    override def toString = segs.mkString(".")
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: This takes the Mongo approach of destroying information on arrays. I think it'd be better to preserve the distinction between index and field, so you can render them differently (e.g. `loc[0]` and `loc[1]`).

File: core/src/main/scala/slamdata/engine/repl/prettify.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    def ::(prefix: String): Path = Path(prefix :: segs)
+  }
+  object Path{
+    def apply(segs: String*): Path = Path(segs.toList)
+  }
+
+
+  def mergePaths(as: List[Path], bs: List[Path]): List[Path] =
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: This has bad performance characteristics -- if this is used for every row in the result set, I suggest instead using `Set`.

File: core/src/main/scala/slamdata/engine/repl/prettify.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      case -\/ (value) => ListMap(Path("value") -> value)
+      case  \/-(map)   => map.toListMap
+    }
+  }
+
+  /**
+   Render any atomic Data value to a String that should either left-aligned (Str values),
+   or right-aligned (all others).
+   */
+  def render(data: Data): String \/ String = data match {
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: Unless you're using the monadic / etc., functionality of `\/`, I'd suggest an ADT here like

```scala
sealed trait Aligned[A]
final case class LeftAligned[A](value: A) extends Alignment[A]
final case class RightAligned[A](value: A) extends Alignment[A]
```

File: core/src/main/scala/slamdata/engine/repl/repl.scala [hunk 4]
```diff
@@ -167,11 +167,13 @@ object Repl {
 
   def select(state: RunState, query: String, name: Option[String]): Process[Task, Unit] =
   {
-    def summarize[A](max: Int)(lines: IndexedSeq[A]): String =
-      if (lines.lengthCompare(0) <= 0) "No results found"
+    val Lines = 10
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/repl/repl.scala]: It'd be nice to make this a variable in the environment. Pulling it out into a constant's a good first step.

File: core/src/main/scala/slamdata/engine/repl/prettify.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    case _ => \/-("unexpected: " + data)  // NB: the non-atomic types never appear here because the Data has been flattened
+  }
+
+  /**
+   Render a list of non-atomic values to a table:
+   - Str values are left-aligned; all other values are right-aligned
+   - There is a column for every field/index, at any depth, that appears in any value.
+   - The table is formatted as in GitHub-flavored MarkDown.
+   */
+  def renderTable(rows: List[Data]): List[String] =
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: Rendering code is hard to test because of all the edge cases, but if you've tested manually on different data sets, happy to merge when you're ready.

**Comment by jdegoes** [comment]
This looks really good, it's going to make a big improvement to usability. Btw, one of the persistent requests we get is to make the table expand to fill the width of the pane. Any chance that's a one-liner?

File: core/src/main/scala/slamdata/engine/repl/prettify.scala (hunk 1 above)
Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: That's a good idea.

File: core/src/main/scala/slamdata/engine/repl/prettify.scala (hunk 2 above)
Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: It needs to preserve insertion order. I'll have to see what Scala has to offer (ala Java's LinkedHashMap/Set).

File: core/src/main/scala/slamdata/engine/repl/prettify.scala (hunk 3 above)
Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: Sure. I had such an ADT in the UI version of this code but when I pulled it out here I ended up with just the two alignment cases and I'm afraid I was seduced by the Left/Right pun.

File: core/src/main/scala/slamdata/engine/repl/repl.scala (hunk 4 above)
Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/repl/repl.scala]: Yes, that was my thinking and if I had a good idea for what to call it I probably would have implemented it. I'm pretty sure some of the traditional tools have such a thing, but I'm not having any luck digging it up in docs for `isql` or `sqlplus`. Any preference.

**Comment by mossprescott** [comment]
Unfortunately, JTable has never made that easy to accomplish (well, you can make it fill the width, but then it refuses to scroll when they don't fit). I think I had suggested auto-sizing the columns, which should be *almost* a one-liner. Let me give that a try.

File: core/src/main/scala/slamdata/engine/repl/repl.scala (hunk 4 above)
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/repl/repl.scala]: Well, in Linux there's the `head` command which is used for this purpose, e.g. `cat foo | head -n 5`. Since variables are used in queries, it's best to pick something not too common like `limit`.

File: core/src/main/scala/slamdata/engine/repl/prettify.scala (hunk 2 above)
Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/repl/prettify.scala]: It looks like the answer is no, Scala doesn't have an immutable, insertion-order preserving, efficient Set. So I'm inclined to leave it as is. This code is only ever applied to a limited number of rows at once, the size of a single batch in the UI (100) or of the summary we show in the REPL (10). And there are worse inefficiencies here, for that matter.

File: admin/src/main/scala/slamdata/engine/admin/table.scala [hunk 6]
```diff
@@ -29,28 +30,33 @@ class CollectionTableModel(fs: FileSystem, path: ResultPath) extends javax.swing
...
 
   def getColumnCount: Int = columns.length max 1
   def getRowCount: Int = collectionSize.getOrElse(1)
+  // NB: in case the table does not use a custom renderer
   def getValueAt(row: Int, column: Int): Object =
+    Styled.render(getData(row, column)).str
+  override def getColumnName(column: Int) =
+    columns.drop(column).headOption.fold("value")(_.label)  // TODO: prune common prefix
+
+  // None == "loading"; Left == missing from result
```
Comment (sellout) [inline admin/src/main/scala/slamdata/engine/admin/table.scala]: Should this have a simple ADT, rather than needing an explanation in a comment?

Comment (jdegoes) [inline admin/src/main/scala/slamdata/engine/admin/table.scala]: :+1: Comments are a code smell. :smiling_imp:

Comment (sellout) [inline admin/src/main/scala/slamdata/engine/admin/table.scala]: Can someone write a macro that gives us Haskellish `data` declarations? 😉

Seriously, though, the verbosity of the Scala version when compared with
```haskell
data Value = Loading | Missing | Value Data
```
makes it really tempting to use existing types.

**Comment by sellout** [comment]
Looks good, just the ADT. I’m excited to have these changes in the REPL.

File: admin/src/main/scala/slamdata/engine/admin/table.scala (hunk 6 above)
Comment (mossprescott) [inline admin/src/main/scala/slamdata/engine/admin/table.scala]: Sigh, yes, there should probably be an ADT there. I *think* I can use the new one in place of one of the other two ADTS I added in this PR, so that's not too horrible.
